pkg/network/java/hotspot.go

```diff
@@ -0,0 +1,262 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+// +build linux
+
+package java
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/fs"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+)
+
+type Hotspot struct {
+	pid   int
+	nsPid int
+	root  string
+	cwd   string // viewed by the process
+	conn  *net.UnixConn
+}
+
+// NewHotspot create an object to connect to an JVM hotspot
+// pid (host pid) and nsPid (within the namespace pid)
+func NewHotspot(pid int, nsPid int) (h *Hotspot, err error) {
+	h = &Hotspot{
+		pid:   pid,
+		nsPid: nsPid,
+	}
+	// workaround Centos 7 (kernel 3.10) NSPid field was introduced in 4.1
+	// so we don't support container on Centos 7
+	if h.nsPid == 0 {
+		h.nsPid = pid
+	}
+
+	procPath := fmt.Sprintf("%s/%d", util.HostProc(), pid)
+	h.root = procPath + "/root"
+	h.cwd, err = os.Readlink(procPath + "/cwd")
+	if err != nil {
+		return nil, err
+	}
+	return h, nil
+}
+
+func (h *Hotspot) tmpPath() string {
+	return fmt.Sprintf("%s/tmp", h.root)
+}
+
+func (h *Hotspot) socketExist() bool {
+	mode, err := os.Stat(fmt.Sprintf("%s/.java_pid%d", h.tmpPath(), h.nsPid))
+	if err == nil && (mode.Mode()&fs.ModeSocket > 0) {
+		return true
+	}
+	return false
+}
+
+func getPathOwner(path string) (uint32, uint32, error) {
+	mode, err := os.Stat(path)
+	if err != nil {
+		return 0, 0, err
+	}
+	stat := mode.Sys().(*syscall.Stat_t)
+	return stat.Uid, stat.Gid, nil
+}
+
+func (h *Hotspot) copyAgent(agent string, uid int, gid int) (dstPath string, cleanup func(), err error) {
+	dstPath = h.cwd + "/" + filepath.Base(agent)
+	if dst, err := os.Stat(h.root + dstPath); err == nil {
+		// if the destination file already exist
+		// check if it's not the source agent file
+		if src, err := os.Stat(agent); err == nil {
+			s := src.Sys().(*syscall.Stat_t)
+			d := dst.Sys().(*syscall.Stat_t)
+			if s.Dev == d.Dev && s.Ino == d.Ino {
+				return "", func() {}, nil
+			}
+		}
+	}
+
+	fagent, err := os.Open(agent)
+	if err != nil {
+		return "", func() {}, err
+	}
+	defer fagent.Close()
+
+	dst, err := os.OpenFile(h.root+dstPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, os.FileMode(0444))
+	if err != nil {
+		return "", func() {}, err
+	}
+	_, err = io.Copy(dst, fagent)
+	if err != nil {
+		return "", func() {}, err
+	}
+	dst.Close()
```

Comment:
use `defer dst.Close()`

Reply:
but if there's error with the copy you are leaking the handle